plasmapy/utils/data/downloader.py

```diff
@@ -4,74 +4,409 @@
+
+    validate : bool, optional
+        If `True`, verify that local files are up-to-date with the data
+        repository, and use the GitHub API to verify download URLs before
+        downloading files. If `False`, return any matching local file without
+        verification and, if a local file cannot be found, attempt to download
+        blindly from the repository. Default is `True`.
+
     """
 
-    if "." not in str(basename):
-        raise ValueError(f"'filename' ({basename}) must include an extension.")
+    # URL for the PlasmaPy-data repository through the GitHub API
+    _API_BASE_URL = "https://api.github.com/repos/PlasmaPy/PlasmaPy-data/contents/"
 
-    if directory is None:
-        directory = Path(Path.home(), ".plasmapy", "downloads")
+    # Name for the local file that stores the SHA hashes of downloaded files
+    # and information about the SHA on the server
+    _blob_file_name = "RESOURCE_BLOB_SHA.json"
 
-        # Create the .plasmapy/downloads directory if it does not already
-        # exist
-        if not directory.is_dir():
-            directory.mkdir()
+    # Base URL for RAW files
+    _RAW_BASE_URL = "https://raw.githubusercontent.com/PlasmaPy/PlasmaPy-data/main/"
 
-    path = Path(directory, basename)
+    def __init__(self, directory: Path | None = None, validate: bool = True):
+        if directory is None:
+            # No test coverage for default directory, since pytest always
+            # saves into a temporary directory
+            self._download_directory = Path(  # coverage: ignore
+                Path.home(),
+                ".plasmapy",
+                "downloads",
+            )
+        else:
+            self._download_directory = Path(directory)
 
-    # If file doesn't exist locally, download it
-    if not path.is_file():
-        url = urljoin(base_url, basename)
+        self._validate = validate
 
-        reply = requests.get(url)  # noqa: S113
+        # Flag to record whether the blob file has been updated from the repo
+        # by this instantiation of the class. Once the file has been updated
+        # once, we won't update it again to limit API calls.
+        self._updated_blob_file_from_repo = False
 
-        if reply.status_code == 404:
-            raise OSError(
-                "The requested URL returned a 404 code, which "
-                "likely indicates that the file does not exist at the "
-                "URL provided."
+        # Make the directory if it doesn't already exist
+        self._download_directory.mkdir(parents=True, exist_ok=True)
+
+        # Path to the local SHA blob file
+        self._blob_file = Path(self._download_directory, self._blob_file_name)
+
+        # Create the SHA blob file if it doesn't already exist
+        if not self._blob_file.is_file():
+            self._blob_dict = {}
+            self._write_blobfile()
+        # Otherwise, read the SHA blob file
+        else:
+            self._read_blobfile()
```

Comment:
I'm wondering if we could remove the need for the comment by renaming `_read_blobfile` to `_read_sha_blobfile`.

Reply:
Yeah, honestly the comment is not really necessary as is